Orange/classification/_tree_scorers.pyx

```diff
@@ -0,0 +1,293 @@
+                best_entro = entro
+                best_idx = i
+    free(distr)
+    return (class_entro - best_entro) / N / log(2), x[idx[best_idx]]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def find_binarization_entropy(np.ndarray[np.float64_t, ndim=2] cont,
+                              np.ndarray[np.float64_t, ndim=1] class_distr,
+                              np.ndarray[np.float64_t, ndim=1] val_distr,
+                              int min_leaf):
+    cdef unsigned int n_classes = cont.shape[0]
+    cdef unsigned int n_values = cont.shape[1]
+    cdef double *distr = <double *>calloc(2 * n_classes, sizeof(double))
+    cdef double *mfrom
+    cdef double *mto
+    cdef double left, right
+    cdef unsigned int i, change, to_right, allowed, m
+    cdef unsigned int best_mapping, move, mapping, previous
+    cdef double entro, class_entro, best_entro
+    cdef double N = 0
+
+    class_entro = 0
+    for i in range(n_classes):
+        distr[i + n_classes] = 0
+        distr[i] = class_distr[i]
+        if class_distr[i] > 0:
+            N += class_distr[i]
+            class_entro -= class_distr[i] * log(class_distr[i])
+    class_entro += N * log(N)
+    best_entro = class_entro
+    left = N
+    right = 0
+
+    previous = 0
+    # Gray code
+    for m in range(1, 1 << (n_values - 1)):
+        # What moves where
+        mapping = m ^ (m >> 1)
+        change = mapping ^ previous
+        to_right = change & mapping
+        for move in range(n_values):
+            if change & 1:
+                break
+            change = change >> 1
+        previous = mapping
+
+        if to_right:
+            left -= val_distr[move]
+            right += val_distr[move]
+            mfrom = distr
+            mto = distr + n_classes
+        else:
+            left += val_distr[move]
+            right -= val_distr[move]
+            mfrom = distr + n_classes
+            mto = distr
+
+        allowed = left >= min_leaf and right >= min_leaf
+        # Move distribution to the other side and
+        # compute entropy by the way, if the split is allowed
+        entro = 0
+        for i in range(n_classes):
+            mfrom[i] -= cont[i, move]
+            mto[i] += cont[i, move]
+            if allowed:
+                if mfrom[i]:
+                    entro -= mfrom[i] * log(mfrom[i])
+                if mto[i]:
+                    entro -= mto[i] * log(mto[i])
+
+        if allowed:
+            entro += left * log(left) + right * log(right)
+            if entro < best_entro:
+                best_entro = entro
+                best_mapping = mapping
+    free(distr)
+    return (class_entro - best_entro) / N / log(2), best_mapping
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def find_threshold_MSE(np.ndarray[np.float64_t, ndim=1] x,
+                       np.ndarray[np.float64_t, ndim=1] y,
+                       np.ndarray[np.int64_t, ndim=1] idx,
+                       int min_leaf):
```

Comment:
Once you have functions use buffers and no python (numpy) code, you can tag them:
```cython
def func(...) nogil:
    ...
```
or
```cython
with nogil:
    ... code that is run with released GIL  ...
```